zerver/tests/test_slack_importer.py

```diff
@@ -0,0 +1,409 @@
+        self.assertEqual(zerver_stream[0]['invite_only'], not channel_data[0]["is_general"])
+        self.assertEqual(zerver_stream[0]['realm'], realm_id)
+        self.assertEqual(zerver_stream[2]['id'],
+                         test_added_channels[zerver_stream[2]['name']])
+
+    @mock.patch("zerver.lib.slack_data_to_zulip_data.get_zerver_realm", return_value=[{}])
+    @mock.patch("zerver.lib.slack_data_to_zulip_data.users_to_zerver_userprofile",
+                return_value=[[], {}])
+    @mock.patch("zerver.lib.slack_data_to_zulip_data.channels_to_zerver_stream",
+                return_value=[[], [], {}, [], [], {}])
+    def test_slack_workspace_to_realm(self, mock_channels_to_zerver_stream: mock.Mock,
+                                      mock_users_to_zerver_userprofile: mock.Mock,
+                                      mock_get_zerver_realm: mock.Mock) -> None:
+
+        realm_id = 1
+        realm, added_users, added_recipient, added_channels = slack_workspace_to_realm(realm_id,
+                                                                                       'test-realm',
+                                                                                       './fixture',
+                                                                                       './random_path')
+        test_zerver_realmdomain = [{'realm': realm_id, 'allow_subdomains': False,
+                                    'domain': 'zulipchat.com', 'id': realm_id}]
+        # Functioning already tests in helper functions
+        self.assertEqual(added_users, {})
+        self.assertEqual(added_channels, {})
+        self.assertEqual(added_recipient, {})
+
+        zerver_realmdomain = realm['zerver_realmdomain']
+        self.assertListEqual(zerver_realmdomain, test_zerver_realmdomain)
+        self.assertEqual(realm['zerver_userpresence'], [])
+        self.assertEqual(realm['zerver_stream'], [])
+        self.assertEqual(realm['zerver_userprofile'], [])
+        self.assertEqual(realm['zerver_realm'], [{}])
+
+    @mock.patch("os.listdir", return_value = ['2015-08-08.json', '2016-01-15.json'])
+    @mock.patch("zerver.lib.slack_data_to_zulip_data.open_data_file")
+    def test_get_total_messages_and_usermessages(self, mock_open_data_file: mock.Mock,
+                                                 mock_list_dir: mock.Mock) -> None:
+
+        date1 = [{"text": "<@U8VAHEVUY> has joined the channel", "subtype": "channel_join"},
+                 {"text": "message"},
+                 {"text": "random"},
+                 {"text": "test messsage"}]
+        date2 = [{"text": "test message 2", "subtype": "channel_leave"},
+                 {"text": "random test"},
+                 {"text": "message", "subtype": "channel_name"}]
+        mock_open_data_file.side_effect = [date1, date2]
+
+        added_recipient = {'random': 2}
+        zerver_subscription = [{'recipient': 2}, {'recipient': 4}, {'recipient': 2}]
```

Comment:
Are the duplicate subscriptions intentional? Are they allowed?

Reply:
A stream would have only one recipient id. Subscription object would check whichever users are subscribed to the stream, so if we have three users subscribed to one stream, we would correspondingly have 3 subscriptions for each user.
This is the creation function: https://github.com/rheaparekh/zulip/blob/df8675376dcb9802f59de7a6860f7281be95f558/zerver/lib/slack_data_to_zulip_data.py#L391-L401. Subscription object is

Reply:
So this is accounting for an object `zerver_subscription` like:
```
                        [{'recipient': 2,
                           'user_profile': 3,
                           'id': 1},
                          {'recipient': 4,
                           'user_profile': 5,
                           'id': 2},
                          {'recipient': 2,
                           'user_profile': 7,
                           'id': 3}]
```

Reply:
(https://github.com/rheaparekh/zulip/blob/df8675376dcb9802f59de7a6860f7281be95f558/zerver/tests/test_slack_importer.py#L163-L164)

Reply:
There are 4 messages in total, but only 8 total user messages, which means the multiplier is 2, so it is as if the last `{'recipient': 2}` is not counted at all because it is a duplicate of the first element.

Reply:
@rht Every subscriptions counts for a user in that stream and user message deals with the message relation to the user. Hence, the multiplier 2 is because of the two users with recipients `2` in `zerver_subscription`.

For `user 1` with recipient `2` in `zerver_subscription`, we will have 4 user_messages corresponding to the 4 messages in that stream. And same would go for `user 2`. Hence total of 8 user_messages

Reply:
Ok, I got tripped over by the fact that 'recipient' is actually stream-specific while the name suggests it is user-specific.

Reply:
@showell my feedback (orthogonal to this PR) is that the naming 'recipient' is confusing to me, where it would have been clearer with 'subscription' corresponding to a subscription id.

Reply:
I agree with the above point,initially I was confused about this too as the name `recipient`is not intuitive here for the zerver_subscription.